File: app/src/main/java/info/nightscout/androidaps/plugins/treatments/fragments/TreatmentsBolusFragment.java
package info.nightscout.androidaps.plugins.treatments.fragments;

import android.graphics.Paint;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.data.Iob;
import info.nightscout.androidaps.data.Profile;
import info.nightscout.androidaps.db.Source;
import info.nightscout.androidaps.events.EventTreatmentChange;
import info.nightscout.androidaps.plugins.bus.RxBus;
import info.nightscout.androidaps.plugins.configBuilder.ProfileFunctions;
import info.nightscout.androidaps.plugins.general.nsclient.NSUpload;
import info.nightscout.androidaps.plugins.general.nsclient.UploadQueue;
import info.nightscout.androidaps.plugins.general.nsclient.events.EventNSClientRestart;
import info.nightscout.androidaps.plugins.iob.iobCobCalculator.events.EventAutosensCalculationFinished;
import info.nightscout.androidaps.plugins.treatments.Treatment;
import info.nightscout.androidaps.plugins.treatments.TreatmentsPlugin;
import info.nightscout.androidaps.plugins.treatments.dialogs.WizardInfoDialog;
import info.nightscout.androidaps.utils.DateUtil;
import info.nightscout.androidaps.utils.DecimalFormatter;
import info.nightscout.androidaps.utils.FabricPrivacy;
import info.nightscout.androidaps.utils.OKDialog;
import info.nightscout.androidaps.utils.SP;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;

import static info.nightscout.androidaps.utils.DateUtil.now;

public class TreatmentsBolusFragment extends Fragment {
    private CompositeDisposable disposable = new CompositeDisposable();

    private RecyclerView recyclerView;

    private TextView iobTotal;
    private TextView activityTotal;
    private Button deleteFutureTreatments;

    public class RecyclerViewAdapter extends RecyclerView.Adapter<RecyclerViewAdapter.TreatmentsViewHolder> {

        List<Treatment> treatments;

        RecyclerViewAdapter(List<Treatment> treatments) {
            this.treatments = treatments;
        }

        @NonNull
        @Override
        public TreatmentsViewHolder onCreateViewHolder(ViewGroup viewGroup, int viewType) {
            View v = LayoutInflater.from(viewGroup.getContext()).inflate(R.layout.treatments_bolus_item, viewGroup, false);
            return new TreatmentsViewHolder(v);
        }

        @Override
        public void onBindViewHolder(@NonNull TreatmentsViewHolder holder, int position) {
            Profile profile = ProfileFunctions.getInstance().getProfile();
            if (profile == null)
                return;
            Treatment t = treatments.get(position);
            holder.date.setText(DateUtil.dateAndTimeString(t.date));
            holder.insulin.setText(MainApp.gs(R.string.formatinsulinunits, t.insulin));
            holder.carbs.setText(MainApp.gs(R.string.format_carbs, (int) t.carbs));
            Iob iob = t.iobCalc(System.currentTimeMillis(), profile.getDia());
            holder.iob.setText(MainApp.gs(R.string.formatinsulinunits, iob.iobContrib));
            holder.mealOrCorrection.setText(t.isSMB ? "SMB" : t.mealBolus ? MainApp.gs(R.string.mealbolus) : MainApp.gs(R.string.correctionbous));
            holder.ph.setVisibility(t.source == Source.PUMP ? View.VISIBLE : View.GONE);
            holder.ns.setVisibility(NSUpload.isIdValid(t._id) ? View.VISIBLE : View.GONE);
            holder.invalid.setVisibility(t.isValid ? View.GONE : View.VISIBLE);
            if (iob.iobContrib != 0)
                holder.iob.setTextColor(ContextCompat.getColor(MainApp.instance(), R.color.colorActive));
            else
                holder.iob.setTextColor(holder.carbs.getCurrentTextColor());
            if (t.date > now())
                holder.date.setTextColor(ContextCompat.getColor(MainApp.instance(), R.color.colorScheduled));
            else
                holder.date.setTextColor(holder.carbs.getCurrentTextColor());
            holder.remove.setTag(t);
            holder.calculation.setTag(t);
            holder.calculation.setVisibility(t.getBoluscalc() == null ? View.INVISIBLE : View.VISIBLE);
        }

        @Override
        public int getItemCount() {
            return treatments.size();
        }

        @Override
        public void onAttachedToRecyclerView(@NonNull RecyclerView recyclerView) {
            super.onAttachedToRecyclerView(recyclerView);
        }

        public class TreatmentsViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
            CardView cv;
            TextView date;
            TextView insulin;
            TextView carbs;
            TextView iob;
            TextView mealOrCorrection;
            TextView remove;
            TextView calculation;
            TextView ph;
            TextView ns;
            TextView invalid;

            TreatmentsViewHolder(View itemView) {
                super(itemView);
                cv = itemView.findViewById(R.id.treatments_cardview);
                date = itemView.findViewById(R.id.treatments_date);
                insulin = itemView.findViewById(R.id.treatments_insulin);
                carbs = itemView.findViewById(R.id.treatments_carbs);
                iob = itemView.findViewById(R.id.treatments_iob);
                mealOrCorrection = itemView.findViewById(R.id.treatments_mealorcorrection);
                ph = itemView.findViewById(R.id.pump_sign);
                ns = itemView.findViewById(R.id.ns_sign);
                invalid = itemView.findViewById(R.id.invalid_sign);
                calculation = itemView.findViewById(R.id.treatments_calculation);
                calculation.setOnClickListener(this);
                calculation.setPaintFlags(calculation.getPaintFlags() | Paint.UNDERLINE_TEXT_FLAG);
                remove = itemView.findViewById(R.id.treatments_remove);
                remove.setOnClickListener(this);
                remove.setPaintFlags(remove.getPaintFlags() | Paint.UNDERLINE_TEXT_FLAG);
            }

            @Override
            public void onClick(View v) {
                final Treatment treatment = (Treatment) v.getTag();
                if (treatment == null)
                    return;
                switch (v.getId()) {
                    case R.id.treatments_remove:
                        OKDialog.showConfirmation(getContext(), MainApp.gs(R.string.removerecord),
                                MainApp.gs(R.string.configbuilder_insulin) + ": " + MainApp.gs(R.string.formatinsulinunits, treatment.insulin) +
                                        "\n" + MainApp.gs(R.string.carbs) + ": " + MainApp.gs(R.string.format_carbs, (int) treatment.carbs) +
                                        "\n" + MainApp.gs(R.string.date) + ": " + DateUtil.dateAndTimeString(treatment.date),
                                (dialog, id) -> {
                                    final String _id = treatment._id;
                                    if (treatment.source == Source.PUMP) {
                                        treatment.isValid = false;
                                        TreatmentsPlugin.getPlugin().getService().update(treatment);
                                    } else {
                                        if (NSUpload.isIdValid(_id)) {
                                            NSUpload.removeCareportalEntryFromNS(_id);
                                        } else {
                                            UploadQueue.removeID("dbAdd", _id);
                                        }
                                        TreatmentsPlugin.getPlugin().getService().delete(treatment);
                                    }
                                    updateGui();
                                }, null);
                        break;
                    case R.id.treatments_calculation:
                        FragmentManager manager = getFragmentManager();
                        // try to fix  https://fabric.io/nightscout3/android/apps/info.nightscout.androidaps/issues/5aca7a1536c7b23527eb4be7?time=last-seven-days
                        // https://stackoverflow.com/questions/14860239/checking-if-state-is-saved-before-committing-a-fragmenttransaction
                        if (manager.isStateSaved())
                            return;
                        if (treatment.getBoluscalc() != null) {
                            WizardInfoDialog wizardDialog = new WizardInfoDialog();
                            wizardDialog.setData(treatment.getBoluscalc());
                            wizardDialog.show(manager, "WizardInfoDialog");
                        }
                        break;
                }
            }
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.treatments_bolus_fragment, container, false);

        recyclerView = view.findViewById(R.id.treatments_recyclerview);
        recyclerView.setHasFixedSize(true);
        LinearLayoutManager llm = new LinearLayoutManager(view.getContext());
        recyclerView.setLayoutManager(llm);

        RecyclerViewAdapter adapter = new RecyclerViewAdapter(TreatmentsPlugin.getPlugin().getTreatmentsFromHistory());
        recyclerView.setAdapter(adapter);

        iobTotal = view.findViewById(R.id.treatments_iobtotal);
        activityTotal = view.findViewById(R.id.treatments_iobactivitytotal);

        Button refreshFromNS = view.findViewById(R.id.treatments_reshreshfromnightscout);
        refreshFromNS.setOnClickListener(v -> OKDialog.showConfirmation(getContext(), MainApp.gs(R.string.refresheventsfromnightscout) + "?", () -> {
            TreatmentsPlugin.getPlugin().getService().resetTreatments();
            RxBus.INSTANCE.send(new EventNSClientRestart());
        }));

        deleteFutureTreatments = view.findViewById(R.id.treatments_delete_future_treatments);
        deleteFutureTreatments.setOnClickListener(v -> {
            OKDialog.showConfirmation(getContext(), MainApp.gs(R.string.overview_treatment_label), MainApp.gs(R.string.deletefuturetreatments) + "?", () -> {
                final List<Treatment> futureTreatments = TreatmentsPlugin.getPlugin().getService()
                        .getTreatmentDataFromTime(now() + 1000, true);
                for (Treatment treatment : futureTreatments) {
                    final String _id = treatment._id;
                    if (NSUpload.isIdValid(_id)) {
                        NSUpload.removeCareportalEntryFromNS(_id);
                    } else {
                        UploadQueue.removeID("dbAdd", _id);
                    }
                    TreatmentsPlugin.getPlugin().getService().delete(treatment);
                }
                updateGui();
            });
        });

        boolean nsUploadOnly = SP.getBoolean(R.string.key_ns_upload_only, true);
        if (nsUploadOnly)
            refreshFromNS.setVisibility(View.GONE);

        return view;
    }

    @Override
    public synchronized void onResume() {
        super.onResume();
        disposable.add(RxBus.INSTANCE
                .toObservable(EventTreatmentChange.class)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(event -> updateGui(), FabricPrivacy::logException)
        );
        disposable.add(RxBus.INSTANCE
                .toObservable(EventAutosensCalculationFinished.class)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(event -> updateGui(), FabricPrivacy::logException)
        );
        updateGui();
    }

    @Override
    public synchronized void onPause() {
        super.onPause();
        disposable.clear();
    }

    private void updateGui() {
        recyclerView.swapAdapter(new RecyclerViewAdapter(TreatmentsPlugin.getPlugin().getTreatmentsFromHistory()), false);
        if (TreatmentsPlugin.getPlugin().getLastCalculationTreatments() != null) {
            iobTotal.setText(DecimalFormatter.to2Decimal(TreatmentsPlugin.getPlugin().getLastCalculationTreatments().iob) + " " + MainApp.gs(R.string.insulin_unit_shortname));
            activityTotal.setText(DecimalFormatter.to3Decimal(TreatmentsPlugin.getPlugin().getLastCalculationTreatments().activity) + " " + MainApp.gs(R.string.insulin_unit_shortname));
        }
        if (!TreatmentsPlugin.getPlugin().getService().getTreatmentDataFromTime(now() + 1000, true).isEmpty()) {
            deleteFutureTreatments.setVisibility(View.VISIBLE);
        } else {
            deleteFutureTreatments.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/info/nightscout/androidaps/plugins/treatments/fragments/TreatmentsBolusFragment.kt
package info.nightscout.androidaps.plugins.treatments.fragments

import android.graphics.Paint
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dagger.android.support.DaggerFragment
import info.nightscout.androidaps.R
import info.nightscout.androidaps.databinding.TreatmentsBolusFragmentBinding
import info.nightscout.androidaps.databinding.TreatmentsBolusItemBinding
import info.nightscout.androidaps.db.Source
import info.nightscout.androidaps.db.Treatment
import info.nightscout.androidaps.dialogs.WizardInfoDialog
import info.nightscout.androidaps.events.EventTreatmentChange
import info.nightscout.androidaps.interfaces.ProfileFunction
import info.nightscout.androidaps.logging.UserEntryLogger
import info.nightscout.androidaps.plugins.bus.RxBusWrapper
import info.nightscout.androidaps.plugins.general.nsclient.NSUpload
import info.nightscout.androidaps.plugins.general.nsclient.UploadQueue
import info.nightscout.androidaps.plugins.general.nsclient.events.EventNSClientRestart
import info.nightscout.androidaps.plugins.iob.iobCobCalculator.events.EventAutosensCalculationFinished
import info.nightscout.androidaps.plugins.treatments.TreatmentsPlugin
import info.nightscout.androidaps.plugins.treatments.fragments.TreatmentsBolusFragment.RecyclerViewAdapter.TreatmentsViewHolder
import info.nightscout.androidaps.utils.DateUtil
import info.nightscout.androidaps.utils.FabricPrivacy
import info.nightscout.androidaps.utils.alertDialogs.OKDialog
import info.nightscout.androidaps.utils.buildHelper.BuildHelper
import info.nightscout.androidaps.utils.resources.ResourceHelper
import info.nightscout.androidaps.utils.rx.AapsSchedulers
import info.nightscout.androidaps.utils.sharedPreferences.SP
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

class TreatmentsBolusFragment : DaggerFragment() {

    private val disposable = CompositeDisposable()

    @Inject lateinit var rxBus: RxBusWrapper
    @Inject lateinit var sp: SP
    @Inject lateinit var resourceHelper: ResourceHelper
    @Inject lateinit var fabricPrivacy: FabricPrivacy
    @Inject lateinit var treatmentsPlugin: TreatmentsPlugin
    @Inject lateinit var profileFunction: ProfileFunction
    @Inject lateinit var nsUpload: NSUpload
    @Inject lateinit var uploadQueue: UploadQueue
    @Inject lateinit var dateUtil: DateUtil
    @Inject lateinit var buildHelper: BuildHelper
    @Inject lateinit var aapsSchedulers: AapsSchedulers
    @Inject lateinit var uel: UserEntryLogger

    private var _binding: TreatmentsBolusFragmentBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View =
        TreatmentsBolusFragmentBinding.inflate(inflater, container, false).also { _binding = it }.root

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.recyclerview.setHasFixedSize(true)
        binding.recyclerview.layoutManager = LinearLayoutManager(view.context)
        binding.recyclerview.adapter = RecyclerViewAdapter(treatmentsPlugin.treatmentsFromHistory)
        binding.refreshFromNightscout.setOnClickListener {
            activity?.let { activity ->
                OKDialog.showConfirmation(activity, resourceHelper.gs(R.string.refresheventsfromnightscout) + "?") {
                    uel.log("TREAT NS REFRESH")
                    treatmentsPlugin.service.resetTreatments()
                    rxBus.send(EventNSClientRestart())
                }
            }
        }
        binding.deleteFutureTreatments.setOnClickListener {
            activity?.let { activity ->
                OKDialog.showConfirmation(activity, resourceHelper.gs(R.string.overview_treatment_label), resourceHelper.gs(R.string.deletefuturetreatments) + "?", Runnable {
                    uel.log("DELETE FUTURE TREATMENTS")
                    val futureTreatments = treatmentsPlugin.service.getTreatmentDataFromTime(DateUtil.now() + 1000, true)
                    for (treatment in futureTreatments) {
                        if (NSUpload.isIdValid(treatment._id))
                            nsUpload.removeCareportalEntryFromNS(treatment._id)
                        else
                            uploadQueue.removeID("dbAdd", treatment._id)
                        treatmentsPlugin.service.delete(treatment)
                    }
                    updateGui()
                })
            }
        }
        val nsUploadOnly = sp.getBoolean(R.string.key_ns_upload_only, true) || !buildHelper.isEngineeringMode()
        if (nsUploadOnly) binding.refreshFromNightscout.visibility = View.GONE
    }

    @Synchronized
    override fun onResume() {
        super.onResume()
        disposable.add(rxBus
            .toObservable(EventTreatmentChange::class.java)
            .observeOn(aapsSchedulers.main)
            .subscribe({ updateGui() }, fabricPrivacy::logException)
        )
        disposable.add(rxBus
            .toObservable(EventAutosensCalculationFinished::class.java)
            .observeOn(aapsSchedulers.main)
            .subscribe({ updateGui() }, fabricPrivacy::logException)
        )
        updateGui()
    }

    @Synchronized
    override fun onPause() {
        super.onPause()
        disposable.clear()
    }

    @Synchronized
    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    inner class RecyclerViewAdapter internal constructor(var treatments: List<Treatment>) : RecyclerView.Adapter<TreatmentsViewHolder>() {

        override fun onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): TreatmentsViewHolder {
            val v = LayoutInflater.from(viewGroup.context).inflate(R.layout.treatments_bolus_item, viewGroup, false)
            return TreatmentsViewHolder(v)
        }

        override fun onBindViewHolder(holder: TreatmentsViewHolder, position: Int) {
            val profile = profileFunction.getProfile() ?: return
            val t = treatments[position]
            holder.binding.date.text = dateUtil.dateAndTimeString(t.date)
            holder.binding.insulin.text = resourceHelper.gs(R.string.formatinsulinunits, t.insulin)
            holder.binding.carbs.text = resourceHelper.gs(R.string.format_carbs, t.carbs.toInt())
            val iob = t.iobCalc(System.currentTimeMillis(), profile.dia)
            holder.binding.iob.text = resourceHelper.gs(R.string.formatinsulinunits, iob.iobContrib)
            holder.binding.mealOrCorrection.text = if (t.isSMB) "SMB" else if (t.mealBolus) resourceHelper.gs(R.string.mealbolus) else resourceHelper.gs(R.string.correctionbous)
            holder.binding.pump.visibility = if (t.source == Source.PUMP) View.VISIBLE else View.GONE
            holder.binding.ns.visibility = if (NSUpload.isIdValid(t._id)) View.VISIBLE else View.GONE
            holder.binding.invalid.visibility = if (t.isValid) View.GONE else View.VISIBLE
            if (iob.iobContrib != 0.0) holder.binding.iob.setTextColor(resourceHelper.gc(R.color.colorActive)) else holder.binding.iob.setTextColor(holder.binding.carbs.currentTextColor)
            if (t.date > DateUtil.now()) holder.binding.date.setTextColor(resourceHelper.gc(R.color.colorScheduled)) else holder.binding.date.setTextColor(holder.binding.carbs.currentTextColor)
            holder.binding.remove.tag = t
            holder.binding.calculation.tag = t
            holder.binding.calculation.visibility = if (t.getBoluscalc() == null) View.INVISIBLE else View.VISIBLE
        }

        override fun getItemCount(): Int {
            return treatments.size
        }

        inner class TreatmentsViewHolder internal constructor(view: View) : RecyclerView.ViewHolder(view) {

            val binding = TreatmentsBolusItemBinding.bind(view)

            init {
                binding.calculation.setOnClickListener {
                    val treatment = it.tag as Treatment
                    if (treatment.getBoluscalc() != null) {
                        val wizardDialog = WizardInfoDialog()
                        wizardDialog.setData(treatment.getBoluscalc()!!)
                        wizardDialog.show(childFragmentManager, "WizardInfoDialog")
                    }
                }
                binding.calculation.paintFlags = binding.calculation.paintFlags or Paint.UNDERLINE_TEXT_FLAG
                binding.remove.setOnClickListener {
                    val treatment = it.tag as Treatment? ?: return@setOnClickListener
                    activity?.let { activity ->
                        val text = resourceHelper.gs(R.string.configbuilder_insulin) + ": " +
                            resourceHelper.gs(R.string.formatinsulinunits, treatment.insulin) + "\n" +
                            resourceHelper.gs(R.string.carbs) + ": " + resourceHelper.gs(R.string.format_carbs, treatment.carbs.toInt()) + "\n" +
                            resourceHelper.gs(R.string.date) + ": " + dateUtil.dateAndTimeString(treatment.date)
                        OKDialog.showConfirmation(activity, resourceHelper.gs(R.string.removerecord), text, Runnable {
                            uel.log("REMOVED TREATMENT", text)
                            if (treatment.source == Source.PUMP) {
                                treatment.isValid = false
                                treatmentsPlugin.service.update(treatment)
                            } else {
                                if (NSUpload.isIdValid(treatment._id))
                                    nsUpload.removeCareportalEntryFromNS(treatment._id)
                                else
                                    uploadQueue.removeID("dbAdd", treatment._id)
                                treatmentsPlugin.service.delete(treatment)
                            }
                            updateGui()
                        })
                    }
                }
                binding.remove.paintFlags = binding.remove.paintFlags or Paint.UNDERLINE_TEXT_FLAG
            }
        }
    }

    private fun updateGui() {
        if (_binding == null) return
        binding.recyclerview.swapAdapter(RecyclerViewAdapter(treatmentsPlugin.treatmentsFromHistory), false)
        if (treatmentsPlugin.lastCalculationTreatments != null) {
            binding.iobTotal.text = resourceHelper.gs(R.string.formatinsulinunits, treatmentsPlugin.lastCalculationTreatments.iob)
            binding.iobActivityTotal.text = resourceHelper.gs(R.string.formatinsulinunits, treatmentsPlugin.lastCalculationTreatments.activity)
        }
        if (treatmentsPlugin.service.getTreatmentDataFromTime(DateUtil.now() + 1000, true).isNotEmpty())
            binding.deleteFutureTreatments.visibility = View.VISIBLE
        else
            binding.deleteFutureTreatments.visibility = View.GONE
    }
}